onsp_co2/models/carbon_factor.py

```diff
@@ -139,6 +141,21 @@ def write(self, vals):
                 )
             )
         return super(CarbonFactor, self).write(vals)
+    
+    def write(self, vals):
+        res = super().write(vals)
+        
+        date_dict = defaultdict(lambda: self.env['carbon.factor.value'])
+        for value in self.value_ids:
+            date_dict[str(value.date)] |= value
+            
+        for date, value_list in date_dict.items():
+            value_carbon_factor_type_ids = value_list.mapped('carbon_factor_type_id').sorted()
+            carbon_factor_type_ids = self.carbon_factor_type_ids.sorted()
+            if value_carbon_factor_type_ids != carbon_factor_type_ids:
+                raise ValidationError('Please enter all the required Carbon Factor Type for the date {0}'.format(date))
```

Comment:
Several problems here :
- The logic should be in another method called in the write
- `date_dict` is not a clear variable name, something like `date_to_values` would be better (that describes your variable & follows the "convention" `KEY_to_VALUE` )
- This doesn't work if `self` is a recordset (write is not a "single" method). In the new method you should add `for factor in self`

Also, are recordset ordered? I'm surprised that you need the sorted

Reply:
This code won't work in the final version, but in the current code you could do this 
```python
if self.carbon_factor_type_ids - value_list.carbon_factor_type_id:
    raise ValidationError('Please enter all the required Carbon Factor Type for the date {0}'.format(date))
```

It will be true if there are remaining type after the substraction